pkg/services/ngalert/store/instance_database.go

```diff
@@ -43,6 +44,46 @@ func (st DBstore) ListAlertInstances(ctx context.Context, cmd *models.ListAlertI
 	return result, err
 }
 
+func (st DBstore) FullSync(ctx context.Context, instances []models.AlertInstance) error {
+	if len(instances) == 0 {
+		return nil
+	}
+	return st.SQLStore.WithTransactionalDbSession(ctx, func(sess *sqlstore.DBSession) error {
+		// First we delete all records from the table
+		if _, err := sess.Exec("DELETE FROM alert_instance"); err != nil {
+			return fmt.Errorf("failed to delete alert_instance table: %w", err)
+		}
+		for _, alertInstance := range instances {
+			if err := models.ValidateAlertInstance(alertInstance); err != nil {
+				st.Logger.Warn("Failed to validate alert instance, skipping", "err", err, "rule_uid", alertInstance.RuleUID)
+				continue
+			}
+			labelTupleJSON, err := alertInstance.Labels.StringKey()
+			if err != nil {
+				st.Logger.Warn("Failed to generate alert instance labels key, skipping", "err", err, "rule_uid", alertInstance.RuleUID)
+				continue
+			}
+			params := append(make([]any, 0), alertInstance.RuleOrgID, alertInstance.RuleUID, labelTupleJSON, alertInstance.LabelsHash, alertInstance.CurrentState, alertInstance.CurrentReason, alertInstance.CurrentStateSince.Unix(), alertInstance.CurrentStateEnd.Unix(), alertInstance.LastEvalTime.Unix())
+
+			upsertSQL := st.SQLStore.GetDialect().UpsertSQL(
```

Comment:
Do we need to do an upsert if we've truncated the table? Can we do a regular insert? Would it be faster?

Reply:
Good catch. I just copied it, will test it tomorrow, it should be faster with inserts.

Reply:
Changed it in https://github.com/grafana/grafana/pull/74998/commits/b09fb7010cea0a5128dab384c889bf721cb9be90, seems to be 50% faster in SQLite.